src/coreclr/jit/lower.cpp

```diff
@@ -2708,10 +2708,17 @@ GenTree* Lowering::OptimizeConstCompare(GenTree* cmp)
 
         if (op2->IsIntegralConst(0) && (op1->gtNext == op2) && (op2->gtNext == cmp) &&
 #ifdef TARGET_XARCH
-            op1->OperIs(GT_AND, GT_OR, GT_XOR, GT_ADD, GT_SUB, GT_NEG))
+            (op1->OperIs(GT_AND, GT_OR, GT_XOR, GT_ADD, GT_SUB, GT_NEG)
+#ifdef FEATURE_HW_INTRINSICS
+             || (op1->OperIs(GT_HWINTRINSIC) &&
+                 emitter::DoesWriteZeroFlag(
+                     HWIntrinsicInfo::lookupIns(op1->AsHWIntrinsic()->GetHWIntrinsicId(), op1->TypeGet())))
```

Comment:
@tannergooding could you please take a look at this when you have time?

My concern is that it appears the second operand to `lookupIns` is "sometimes" "the base type" (presumably for SIMD intrinsics) and "sometimes" "the target type" (presumably for scalar intrinsics), and we're always passing "the target type" here (the node type). This feels prone to picking up wrong instructions, and I am not sure of the correct guard against it.

Reply:
Rather than taking the instruction, have an overload that takes the intrinsic node. The method can then use the node and do the relevant switch of "for scalar intrinsics, there is no `simdBaseType` so use the node type; for simd intrinsics use the simdBaseType"

Reply:
Done. Added a lookupIns overload that takes the intrinsic node looks up the category and for scalar uses the node type otherwise gets the base simd type and then looks up the instruction.